lib/watir-webdriver/checkers.rb

```diff
@@ -0,0 +1,88 @@
+module Watir
+  class Checkers
+
+    def initialize(browser)
+      @browser = browser
+      @error_checkers = []
+    end
+
+    #
+    # Adds new checker.
+    #
+    # Checkers are generally used to ensure application under test does not encounter
+    # any error. They are automatically executed after following events:
+    #   1. Open URL.
+    #   2. Refresh page.
+    #   3. Click, double-click or right-click on element.
+    #
+    # @example
+    #   browser.checkers.add do |page|
+    #     page.text.include?("Server Error") and puts "Application exception or 500 error!"
+    #   end
+    #   browser.goto "www.watir.com/404"
+    #   "Application exception or 500 error!"
+    #
+    # @param [#call] checker Object responding to call
+    # @yield Checker block
+    # @yieldparam [Watir::Browser]
+    #
+
+    def add(checker = nil, &block)
+      if block_given?
+        @error_checkers << block
+      elsif checker.respond_to? :call
+        @error_checkers << checker
+      else
+        raise ArgumentError, "expected block or object responding to #call"
+      end
+    end
+
+    #
+    # Deletes checker.
+    #
+    # @example
+    #   checker = lambda do |page|
+    #     page.text.include?("Server Error") and puts "Application exception or 500 error!"
+    #   end
+    #   browser.checkers.add checker
+    #   browser.goto "www.watir.com/404"
+    #   "Application exception or 500 error!"
+    #   browser.disable_checker checker
+    #   browser.refresh
+    #
+
+    def disable(checker)
+      @error_checkers.delete(checker)
+    end
+
+    #
+    # Runs checkers.
+    #
+
+    def run
+      if @error_checkers.any? && @browser.window.present?
```

Comment:
It's for when you have a second window open and you click a button that closes the window. The checkers kick off immediately after the click, but before the driver is gets pointed back to an active window. Without checking for window.present?, anything that makes a wire call will fail because there is nothing to make a call on.

When you're running checkers, alerts and multiple windows are a huge pain (especially since FF & Chrome behave differently), which is why I implemented #without_checkers. Neither of these checks here would be necessary if we instead had methods for #click_for_alert and #click_for_window_open and #click_for_window_close.